signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+        """
+        Return true if input is a non-empty list of strings. Otherwise
+        return false.
+        """
+        return (
+            bool(names)
+            and isinstance(names, list)
+            and all(isinstance(event, str) for event in names)
+        )
+
+    def publish_signal(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> int:
+        """
+        Publish a legal event signal. Includes additional options to specify sample_no,
+        a tag, and any other additional metadata. Will then wait for responses from
+        subscribed responders (if any). Returns one of the below result codes based on 
+        signal publish/response success.
+
+        RESULT CODES:
+        0 = ALL SUBS RESPONDED WELL
+        1 = SOME SUBS RESPONDED BADLY
+        2 = NOT ALL SUBS RESPONDED
+        3 = ILLEGAL EVENT NAME PASSED IN
+        4 = INITIALIZATION SIGNAL ATTEMPTED, BAD
+        5 = SHUTDOWN SIGNAL ATTEMPTED, BAD
+        """
+        skip_check = False
+        if not self.init_listener or not self.init_listener.is_alive():
+            print(
+                "WARNING: Exporter is not initialized, not accepting subscribers and no event checking"
+            )
+            skip_check = True
+
+        if event == "initialization":
+            print(
+                "ERROR: Please use the 'initialize()' method for publishing initialization signals"
+            )
+            return 4
+
+        if event == "shutdown":
+            print("ERROR: Please use the 'shutdown()' method for shutdown signals")
+            return 5
+
+        if not skip_check and not event in self.legal_events:
+            print(f"Event {self.event} not one of legal events: {self.legal_events}")
+            return 3
+
+        sig = self._sig_builder(event=event, sample=sample, tag=tag, metadata=metadata)
+        sub_check, result_box = self._check_subs(event)
+
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+        counter = 0
+        while sub_check and sub_check.is_alive():
+            time.sleep(0.1)
+            counter += 1
+            if counter >= 200:
+                print("Timeout after waiting 20 seconds for sub response")
+                return 2
+
+        return result_box[0]
+
+    def initialize(
+        self, legal_events: list, tag: str = None, expected_hosts: list = None
+    ) -> None:
+        """
+        Publishes an initialization message. Starts a listener that reads responses
+        to the initialization message and adds responders to the subscriber list.
+        Sets list of legal event names for future signals, and also allows for optional
+        input of expected hostnames (subscribers) as well as a tag.
+        """
+        if not self._valid_str_list(legal_events):
+            print("ERROR: 'legal_events' arg must be a list of string event names")
+            return
```

Comment:
Can we raise errors here rather than just return ``None``? There is nothing to differentiate failure in this function vs success.